packages/frontend/core/src/components/attachment-viewer/viewer.tsx

```diff
@@ -0,0 +1,425 @@
+    startIndex: 0,
+    endIndex: 0,
+  });
+
+  const [collapsed, setCollapsed] = useState(true);
+  const thumbnailsScrollerHandleRef = useRef<VirtuosoHandle | null>(null);
+  const thumbnailsScrollerRef = useRef<HTMLElement | null>(null);
+  const [thumbnailsVisibleRange, setThumbnailsVisibleRange] = useState({
+    startIndex: 0,
+    endIndex: 0,
+  });
+
+  const post = useCallback(
+    <T extends MessageOp>(
+      type: T,
+      data?: MessageDataType[T],
+      transfers?: Transferable[]
+    ) => {
+      const message = { type, [type]: data };
+      if (transfers?.length) {
+        workerRef.current?.postMessage(message, transfers);
+        return;
+      }
+      workerRef.current?.postMessage(message);
+    },
+    [workerRef]
+  );
+
+  const render = useCallback(
+    (id: number, kind: RenderKind, imageData: ImageData) => {
+      renderItem(
+        (kind === RenderKind.Page ? scrollerRef : thumbnailsScrollerRef)
+          .current,
+        id,
+        imageData
+      );
+    },
+    [scrollerRef, thumbnailsScrollerRef]
+  );
+
+  const onScroll = useCallback(() => {
+    const el = scrollerRef.current;
+    if (!el) return;
+
+    const { total } = docInfo;
+    if (!total) return;
+
+    const { scrollTop, scrollHeight } = el;
+    const itemHeight = scrollHeight / total;
+    const n = scrollTop / itemHeight;
+    const t = n / total;
+    const index = Math.floor(n + t);
+    const cursor = Math.min(index, total - 1);
+
+    setCursor(cursor);
+  }, [scrollerRef, docInfo]);
+
+  const onSelect = useCallback(
+    (index: number) => {
+      scrollerHandleRef.current?.scrollToIndex({
+        index,
+        align: 'start',
+        behavior: 'smooth',
+      });
+    },
+    [scrollerHandleRef]
+  );
+
+  const updateMainVisibleRange = useMemo(
+    () => debounce(setMainVisibleRange, 233, { trailing: true }),
+    [setMainVisibleRange]
+  );
+
+  const updateThumbnailsVisibleRange = useMemo(
+    () => debounce(setThumbnailsVisibleRange, 233, { trailing: true }),
+    [setThumbnailsVisibleRange]
+  );
+
+  useEffect(() => {
+    const el = viewerRef.current;
+    if (!el) return;
+
+    return observeResize(el, entry => {
+      const rect = entry.contentRect;
+      setViewportInfo(info => ({
+        ...info,
+        width: rect.width,
+        height: rect.height,
+      }));
+    });
+  }, [viewerRef]);
+
+  useEffect(() => {
+    post(MessageOp.Render, {
+      range: mainVisibleRange,
+      kind: RenderKind.Page,
+      scale: 1 * viewportInfo.dpi,
+    });
+  }, [viewportInfo, mainVisibleRange, post]);
+
+  useEffect(() => {
+    if (collapsed) return;
+
+    post(MessageOp.Render, {
+      range: thumbnailsVisibleRange,
+      kind: RenderKind.Thumbnail,
+      scale: (THUMBNAIL_WIDTH / docInfo.width) * viewportInfo.dpi,
+    });
+  }, [collapsed, docInfo, viewportInfo, thumbnailsVisibleRange, post]);
+
+  useLayoutEffect(() => {
+    workerRef.current = new Worker(
```

Comment:
remove from component to make it sharable

Reply:
this part of the reconstruction has been launched in the PR of pdf embed view